### PR TITLE
Lookup: Loop focus in dropdown overlay

### PR DESCRIPTION
**Pull request by marker-dao**

### CODE REVIEW

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js [hunk 1]
```diff
@@ -2167,13 +2172,13 @@ QUnit.module('popup options', {
         const $overlay = $(toSelector(OVERLAY_CONTENT_CLASS)).eq(0);
 
         $(document).trigger('dxpointerdown');
-        assert.equal($overlay.is(':visible'), true, 'overlay is not hidden');
+        assert.equal($overlay.is(':visible'), false, 'overlay is hidden');
     });
 
-    QUnit.test('lookup popup should be hidden after click outside was present', function(assert) {
+    QUnit.test('lookup popup should not be hidden after click outside was present', function(assert) {
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js]:
```suggestion
    QUnit.test('lookup popup should not be hidden after click outside was present if dropDownOptions.hideOnOutsideClick is set to false', function(assert) {
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js [hunk 2]
```diff
@@ -2186,7 +2191,23 @@ QUnit.module('popup options', {
...
+
+    [true, false].forEach(usePopover => {
+        QUnit.test(`Popup should have correct _loopFocus option value if usePopover=${usePopover}`, function(assert) {
+            const instance = $('#lookupOptions').dxLookup({
+                usePopover,
+            }).dxLookup('instance');
+
+            openPopupWithList(instance);
+
+            const $overlay = $(`.${OVERLAY_CLASS}.${POPUP_CLASS}`);
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js]: [possibly excess change]
So won't it work without adding of an overlay class?

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js [hunk 3]
```diff
@@ -2186,7 +2191,23 @@ QUnit.module('popup options', {
...
+            const popup = usePopover ? PopoverFull.getInstance($overlay) : PopupFull.getInstance($overlay);
+            const { _loopFocus } = popup.option();
+
+            assert.equal(_loopFocus, !usePopover);
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js]: [minor]
```suggestion
            assert.strictEqual(_loopFocus, !usePopover, "_loopFocus is correct");
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js [hunk 4]
```diff
@@ -2906,6 +2928,50 @@ QUnit.module('keyboard navigation', {
...
+        const $overlayContent = $(instance.content()).parent();
+        const $cancelButton = $overlayContent.find(CANCEL_BUTTON_SELECTOR);
+
+        $cancelButton.trigger(tabKeyDownEvent);
+
+        assert.ok($element.hasClass(FOCUSED_CLASS), 'lookup field is focused');
+        assert.ok(instance.option('opened'), 'popup is opened');
+    });
+
+    QUnit.test('focus from first Popover element should move back to Lookup field while keeping Popup open when usePopover: true', function(assert) {
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js]:
```suggestion
    QUnit.test('focus from first Popover element should move back to Lookup field while keeping Popup open when usePopover: true' and shift+Tab is pressed, function(assert) {
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js [hunk 5]
```diff
@@ -2906,6 +2928,50 @@ QUnit.module('keyboard navigation', {
...
+            return;
+        }
+
+        const $element = $('#widget').dxLookup({
+            opened: true,
+            items: [1, 2, 3],
+            focusStateEnabled: true,
+        });
+        const instance = $element.dxLookup('instance');
+        const tabKeyDownEvent = $.Event('keydown', { key: 'Tab', shiftKey: true });
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js]:
```suggestion
        const shiftTabKeyDownEvent = $.Event('keydown', { key: 'Tab', shiftKey: true });
```

File: packages/devextreme/js/__internal/ui/overlay/m_overlay.ts [hunk 6]
```diff
@@ -1158,6 +1163,7 @@ const Overlay: typeof OverlayInstance = Widget.inherit({
     switch (name) {
       case 'animation':
         break;
+      case '_loopFocus':
```
Comment (ksercs) [inline packages/devextreme/js/__internal/ui/overlay/m_overlay.ts:1166]: This property runtime change is supposed to be handled more explicit specific way. 
But I agree to leave it as is until this property is private and narrowly used

File: packages/devextreme/js/__internal/ui/m_lookup.ts [hunk 7]
```diff
@@ -573,10 +580,12 @@ const Lookup = DropDownList.inherit({
         hideOnParentScroll: true,
         _fixWrapperPosition: false,
         width: this._isInitialOptionValue('dropDownOptions.width')
-          ? function () { return getOuterWidth(this.$element()); }.bind(this)
-          : this._popupConfig().width,
+          ? () => getOuterWidth(this.$element())
+          : popupConfig.width,
```
Comment (ksercs) [inline packages/devextreme/js/__internal/ui/m_lookup.ts:584]: [possible regression]
are u sure that getting popupConfig once on popover rendering and every time when width function is executed its the same?

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/overlay.tests.js [hunk 8]
```diff
@@ -3276,6 +3276,61 @@ testModule('focus policy', {
...
+        $(lastFocusableElement).focus();
+        $(lastFocusableElement).trigger(this.tabEvent);
+
+        assert.strictEqual(getActiveElement(), firstFocusableElement, 'first item is focused');
+
+        $(firstFocusableElement).trigger(this.shiftTabEvent);
+
+        assert.strictEqual(getActiveElement(), lastFocusableElement, 'last item is focused');
+    });
+
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/overlay.tests.js:3333]: [missed scenario]
In Material lookup when `usePopover=true` now we have looping between input and popup although input is covered with an opened popup :(

File: packages/devextreme/js/__internal/ui/m_lookup.ts (hunk 7 above)
Comment (marker-dao) [reply packages/devextreme/js/__internal/ui/m_lookup.ts:584]: This expression is not a function if this._isInitialOptionValue('dropDownOptions.width') gives false. This means that the right side of the expression is evaluated once when the _renderPopover function is run. That's why this code can be considered safe. And no test has fallen)

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js (hunk 1 above)
Comment (marker-dao) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js]: [c8f7061bdd9cca0c0bb8ba6ec482fd176a987526]

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js (hunk 3 above)
Comment (marker-dao) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js]: [c8f7061bdd9cca0c0bb8ba6ec482fd176a987526]

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js (hunk 4 above)
Comment (marker-dao) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js]: [c8f7061bdd9cca0c0bb8ba6ec482fd176a987526]

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js (hunk 5 above)
Comment (marker-dao) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js]: [c8f7061bdd9cca0c0bb8ba6ec482fd176a987526]

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js (hunk 2 above)
Comment (marker-dao) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/lookup.tests.js]: [dfbacbd2aa478258d57c5229411fccba0d67e7ae]

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/overlay.tests.js (hunk 8 above)
Comment (marker-dao) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/overlay.tests.js:3333]: [8cb97890dfbd36d70e88de459017e78d099c0f2a], [fbceb05cbe4e7e89506dbd78d2f6f670fd68114e], [8bc4fd5f9a822a51660bdda8c11ddf41f82f286c], [1ea569358ffbdf7c72fcc442b079b04afa781a69]

**Review by Zedwag — DISMISSED** [review]

**Review by Zedwag — APPROVED** [review]

**Review by ksercs — APPROVED** [review]

**Pull request merged**